src/Simulation/QsharpCore/Diagnostics/AssertAllZero.qs

```diff
@@ -14,7 +14,7 @@ namespace Microsoft.Quantum.Diagnostics {
     ///
     /// # See Also
     /// - AssertQubit
-    operation AssertAllZero (qubits : Qubit[]) : Unit {
+    operation AssertAllZero (qubits : Qubit[]) : Unit is Adj + Ctl{
```

Comment:
Hm, while the controlled version could be meaningfully defined, I believe supporting it would require additional support in the C++ simulator. We need some tests for this change.

Reply:
As it stands, the controlled version below just ignores the controls, which doesn't seem to match any behavior we'd want. I didn't add that implementation, it already existed here, so I just added the `Adj + Ctl` characteristics for consistency (and to match our language guide recommendation of always including the characteristics). So really, I'm not changing anything here, and the question of if we want a controlled `AssertAllZero` and how it should behave would be a breaking change that we'd have to consider separately from this change.

Reply:
Note that the same is true below for `AssertAllZeroWithinTolerance`. I can't comment on those lines directly, since they didn't change, and the diff on GitHub doesn't show them directly, so I'll include it here for better discoverability:
```qsharp
    operation AssertAllZero (qubits : Qubit[]) : Unit is Adj + Ctl{
        body (...) {
            for (qubit in qubits) {
                AssertQubit(Zero, qubit);
            }
        }

        adjoint self;
        controlled (ctrls, ...) {
            AssertAllZero(qubits);
        }
        controlled adjoint self;
    }
```

Reply:
This seems like a breaking change we might want to consider, but definitively unrelated to this PR. @cgranade I'd like to put this on your plate for API review.